### PR TITLE
Use pthread_join from musl

### PR DESCRIPTION
**Pull request by kleisauke**
Split from PR #13007.

### CODE REVIEW

File: system/lib/libc/musl/src/thread/pthread_cancel.c [hunk 1]
```diff
@@ -3,7 +3,11 @@
 #include "pthread_impl.h"
 #include "syscall.h"
 
+#ifdef __EMSCRIPTEN__
+hidden long __cancel();
+#else
 hidden long __cancel(), __syscall_cp_asm(), __syscall_cp_c();
+#endif
```
Comment (sbc100) [inline system/lib/libc/musl/src/thread/pthread_cancel.c:10]: I don't think this patch should be needed .. declaring extra stuff is petty harmless.

File: system/lib/libc/musl/src/thread/pthread_detach.c [hunk 2]
```diff
@@ -3,10 +3,11 @@
...
 static int __pthread_detach(pthread_t t)
 {
-	// XXX EMSCRIPTEN: Add check for invalid (already joined) thread.  Again
-	// for the benefit of the conformance tests.
-	if (t->self != t)
-		return ESRCH;
+#ifdef __EMSCRIPTEN__
+	// Attempt to detach a thread which does not point to a valid thread, or
+	// does not exist anymore.
+	if (!t || t->self != t) return ESRCH;
```
Comment (sbc100) [inline system/lib/libc/musl/src/thread/pthread_detach.c:9]: Is the test for NULL here needed somewhere?   Most other pthread function don't have this kind of check IIRC.

If it is needed perhaps this can be its own change?

File: system/lib/libc/musl/src/thread/pthread_cond_timedwait.c [hunk 3]
```diff
@@ -187,10 +187,6 @@ int __pthread_cond_timedwait(pthread_cond_t *restrict c, pthread_mutex_t *restri
 		__pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, 0);
 	}
 
-#ifdef __EMSCRIPTEN__
-	pthread_testcancel();
```
Comment (sbc100) [inline system/lib/libc/musl/src/thread/pthread_cond_timedwait.c:191]: Is this needed, or just generally part of a cleanup?  If it not needed as part of this change can you split it out into its own change?

File: system/lib/libc/musl/src/thread/pthread_cancel.c (hunk 1 above)
Comment (kleisauke) [reply system/lib/libc/musl/src/thread/pthread_cancel.c]: Reverted that patch with commit https://github.com/emscripten-core/emscripten/pull/15603/commits/73ed3d4a60d7278e8d277d8e860a4d9293366802 in PR #15603.

File: system/lib/libc/musl/src/thread/pthread_detach.c (hunk 2 above)
Comment (kleisauke) [reply system/lib/libc/musl/src/thread/pthread_detach.c]: The reason for this change was that the previous JS code did check for this, see e.g:
https://github.com/emscripten-core/emscripten/blob/85d58fff45cf47de8df3df2925c6c64e4281b6f7/src/library_pthread.js#L906-L909

I removed this NULL check with commit https://github.com/emscripten-core/emscripten/pull/15604/commits/80a3e1108c81797aa42fca835adcf0d2db9734a0 as it does indeed looks redundant.

File: system/lib/libc/musl/src/thread/pthread_cond_timedwait.c (hunk 3 above)
Comment (kleisauke) [reply system/lib/libc/musl/src/thread/pthread_cond_timedwait.c]: See https://github.com/emscripten-core/emscripten/pull/15603#discussion_r755862451 for the reason for this change, I just moved this to PR #15657.

File: system/lib/libc/musl/src/thread/pthread_cancel.c (hunk 1 above)
Comment (kleisauke) [reply system/lib/libc/musl/src/thread/pthread_cancel.c]: BTW, GitHub does not allow me to stack PRs when I use my own fork to base my branch on (instead of pushing the branches to the main repository). Hence, this PR is currently easiest to review per commit.

File: system/lib/libc/musl/src/thread/pthread_detach.c [hunk 4]
```diff
@@ -3,10 +3,11 @@
 
 static int __pthread_detach(pthread_t t)
 {
-	// XXX EMSCRIPTEN: Add check for invalid (already joined) thread.  Again
-	// for the benefit of the conformance tests.
-	if (t->self != t)
-		return ESRCH;
+#ifdef __EMSCRIPTEN__
+	// Attempt to detach a thread which does not point to a valid thread, or
+	// does not exist anymore.
```
Comment (sbc100) [inline system/lib/libc/musl/src/thread/pthread_detach.c:8]: Why not keep the old comment.. its seems more useful in that it says why we are doing it.  Is it not true?   

The new comment does say why emscripten needs this patch..

File: tests/pthread/main_thread_join.cpp [hunk 5]
```diff
@@ -32,8 +33,12 @@ void *ThreadMain(void *arg) {
   // Delay to force the main thread to try and fail a few times before
   // succeeding.
   while (tries.load() < EXPECTED_TRIES) {}
+#else
+  // Delay a bit to force the main thread to actually block, as exited
+  // threads can be joined without blocking.
+  emscripten_thread_sleep(1000);
```
Comment (sbc100) [inline tests/pthread/main_thread_join.cpp:39]: Is this change actually needed?  What happens if you don't make this change?  Does the test fail?

**Comment by sbc100** [comment]
I wonder if we could split the testcancel change into its own PR?   It seems unrelated to join, and a lot more straight forward, unless I'm missing something?

**Review by sbc100 — COMMENTED** [review]
In general I'm excited to see this land, but maybe we can split it one more time

**Comment by kleisauke** [comment]
> I wonder if we could split the testcancel change into its own PR? It seems unrelated to join, and a lot more straight forward, unless I'm missing something?

Ah, it looks like it can independently land now. I reverted the `pthread_testcancel` with commit abdcdcf9d21d13ebb7e40c0e7595d819ac76525d and will open a new PR with that change.

File: system/lib/libc/musl/src/thread/pthread_detach.c (hunk 4 above)
Comment (kleisauke) [reply system/lib/libc/musl/src/thread/pthread_detach.c]: The reason for the doc comment change was that the "(already joined)"-sentence wasn't quite right. The check is there only for invalid or non-existent threads (AFAIK).

I reverted it partially with commit 1e1761b09811a71c4edd85f956193999d1dc098c, so that it indicates why the patch is needed.

File: tests/pthread/main_thread_join.cpp (hunk 5 above)
Comment (kleisauke) [reply tests/pthread/main_thread_join.cpp:39]: As far as I could remember, the test did indeed fail, because exited threads can be joined without any blocking now. Let me re-check that.

File: system/lib/libc/musl/src/thread/pthread_detach.c (hunk 4 above)
Comment (kleisauke) [reply system/lib/libc/musl/src/thread/pthread_detach.c]: Oh BTW, sometimes there are 2 spaces after a sentence, I've seen that a few times now in comments. Do I have to follow that convention?

Comment (sbc100) [reply system/lib/libc/musl/src/thread/pthread_detach.c]: But I believe the check for validity is to protect specifically again detaching already joined thread (which the testsuite does IIRC).    So its probably worth continuing to mention why we might care specifically about a validity here, even though its undefined behaviour and we technically should not need to care.

File: tests/pthread/main_thread_join.cpp (hunk 5 above)
Comment (sbc100) [reply tests/pthread/main_thread_join.cpp:39]: I believe we should be calling `emscripten_check_blocking_allowed` even when a thread can be joined without blocking.  The idea is to warn folks, even in the cases when they got lucky with the timing.

Comment (sbc100) [reply tests/pthread/main_thread_join.cpp:39]: i.e. the warning/error should still show even when joining already exited threads.

File: system/lib/libc/musl/src/thread/pthread_detach.c (hunk 4 above)
Comment (kleisauke) [reply system/lib/libc/musl/src/thread/pthread_detach.c]: Ah, I understand now. Commit 7806443bae105e1fec61c0c4b1092aa686bbb3b8 brings the comment to its original state. I was misinterpreting that comment as "already-detached", which is checked below.
https://github.com/emscripten-core/emscripten/blob/1e1761b09811a71c4edd85f956193999d1dc098c/system/lib/libc/musl/src/thread/pthread_detach.c#L11-L12

Already joined threads are just threads that are invalid (non-existent).

File: system/lib/libc/musl/src/thread/pthread_join.c [hunk 6]
```diff
@@ -7,26 +7,59 @@ static void dummy1(pthread_t t)
...
+#endif
+
 static int __pthread_timedjoin_np(pthread_t t, void **res, const struct timespec *at)
 {
+#ifdef __EMSCRIPTEN__
+	// Attempt to join a thread which does not point to a valid thread, or
+	// does not exist anymore.
+	if (t->self != t) return ESRCH;
+	// Thread is attempting to join to itself.  Already detached threads are
+	// handled below by returning EINVAL instead.
```
Comment (sbc100) [inline system/lib/libc/musl/src/thread/pthread_join.c:21]: Maybe mention that these extra checks are because we want to pass the posixtest suite?   IIRC both these cases count as undefined behaviour so are not strictly necessary.

**Review by sbc100 — APPROVED** [review]
Nice!  Thanks for your patience on this one.   I'm excited to see this land.

File: tests/pthread/main_thread_join.cpp (hunk 5 above)
Comment (kleisauke) [reply tests/pthread/main_thread_join.cpp]: Reverted that with commit 44776e80a95f40770da7b4ee5f6d55f75c84f092. PR description has been updated accordingly.

File: system/lib/libc/musl/src/thread/pthread_join.c (hunk 6 above)
Comment (kleisauke) [reply system/lib/libc/musl/src/thread/pthread_join.c:21]: Do you mean the extra `t->detach_state != DT_DETACHED` check? AFAIK, that was needed for our `test_main_pthread_join_detach` and `test_proxy_pthread_join_detach` tests, and not needed for the posix tests.

I could also remove this check and update that test exception by partially reverting commit ca90a4b17f8a483474b8e6df2a441febf0413f88, if you want.

Comment (sbc100) [reply system/lib/libc/musl/src/thread/pthread_join.c:21]: Lets leave that as a possible followup.

Comment (kleisauke) [reply system/lib/libc/musl/src/thread/pthread_join.c:21]: I can confirm that the `t->detach_state != DT_DETACHED` check is only needed for the `other.test_{proxy,main}_pthread_join_detach` tests, see for e.g. work-in-progress commit 493f2e510ba37e6c95725ded9bc0db9637fe90ef that updates its test asserts.

**Comment by kleisauke** [comment]
I opened issue https://github.com/emscripten-core/emscripten/issues/16136 to investigate the flakiness of the `browser.test_pthread_run_on_main_thread` test (which also occurs on this PR).

**Review by sbc100 — COMMENTED** [review]
Is the change ready to go?

Regarding the flaky test.. does this change make it more flaky?

File: system/lib/libc/musl/src/thread/pthread_detach.c [hunk 7]
```diff
@@ -16,5 +17,6 @@ static int __pthread_detach(pthread_t t)
 
 weak_alias(__pthread_detach, pthread_detach);
 weak_alias(__pthread_detach, thrd_detach);
-// XXX EMSCRIPTEN: add extra alias for asan.
+#ifdef __EMSCRIPTEN__ // XXX Emscripten add an extra alias for ASan/LSan.
 weak_alias(__pthread_detach, emscripten_builtin_pthread_detach);
+#endif
```
[the remark this answers is not in the document]
Comment (kleisauke) [reply system/lib/libc/musl/src/thread/pthread_detach.c]: 5e65f91c1668cc0bb43598cad2a0717fac5eab1c, I'll prepare a PR that unifies all these `ifdef`'s in musl.

File: system/lib/libc/musl/src/thread/pthread_join.c (hunk 6 above)
Comment (kleisauke) [reply system/lib/libc/musl/src/thread/pthread_join.c:21]: I added a TODO with commit 5e65f91c1668cc0bb43598cad2a0717fac5eab1c, since removing this check would also mean that it causes an `EDEADLK` error when already-detached threads are attempting to detach itself (since `pthread_detach` re-uses code from `pthread_join`).

**Comment by kleisauke** [comment]
> Is the change ready to go?

I think it's ready to land.

> Regarding the flaky test.. does this change make it more flaky?

I'm not sure about this, I didn't fully investigate that flakiness. All tests succeeded prior to the rebase (see commit 898c6131c862f170a605691766de7fe11ccb4ed3), let's see if it passes now.
